### PR TITLE
feat: support cpu inference, distributed inference and binpack/spread schedule policy

### PR DESCRIPTION
**Pull request by aiwantaozi**
https://github.com/gpustack/gpustack/issues/99
https://github.com/gpustack/gpustack/issues/141
https://github.com/gpustack/gpustack/issues/118
https://github.com/gpustack/gpustack/issues/117

### CODE REVIEW

File: gpustack/schemas/models.py [hunk 1]
```diff
@@ -110,11 +136,14 @@ class ModelInstanceBase(SQLModel, ModelSource):
     computed_resource_claim: Optional[ComputedResourceClaim] = Field(
         sa_column=Column(pydantic_column_type(ComputedResourceClaim)), default=None
     )
-    gpu_index: Optional[int] = None
+    gpu_indexes: Optional[List[int]] = Field(sa_column=Column(JSON), default=[])
 
     model_id: int = Field(default=None, foreign_key="models.id")
     model_name: str
 
+    distributes_servers: Optional[DistributedServers] = Field(
```
Comment (gitlawr) [inline gpustack/schemas/models.py]: s/distributes/distributed/

File: gpustack/scheduler/calculator.py [hunk 2]
```diff
@@ -15,19 +16,27 @@
...
+    Partial = "partial"
+    Disable = "disable"
+
+
 @dataclass_json
 @dataclass
-class memoryResource:
-    ram: int
-    vram: int
+class umaVsNonUMA:
```
Comment (gitlawr) [inline gpustack/scheduler/calculator.py]: This name is confusing

File: gpustack/scheduler/scheduler.py [hunk 3]
```diff
@@ -38,6 +47,11 @@ def __init__(self, cfg: Config, check_interval: int = 30):
         self._check_interval = check_interval
         self._engine = get_engine()
         self._queue = AsyncUniqueQueue()
+        self._cache_dir = None
+
+        if self._config.data_dir is not None:
+            self._cache_dir = os.path.join(self._config.data_dir, "parser")
```
Comment (gitlawr) [inline gpustack/scheduler/scheduler.py]: What about `data_dir/cache/gguf-parser`

File: gpustack/scheduler/status_policy.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+from gpustack.schemas.workers import WorkerStateEnum
+
+MaxScore = 100
+
+
+class StatusPolicy:
+    def __init__(self):
+        pass
+
+    async def score(
```
Comment (gitlawr) [inline gpustack/scheduler/status_policy.py]: Why not implement as a filter policy?

File: gpustack/utils/command.py [hunk 5]
```diff
@@ -16,18 +16,12 @@ def is_command_available(command_name):
     return shutil.which(command_name) is not None
 
 
-def get_platform_command(command_map: dict) -> str:
+def get_platform_command(command_map: dict, compute_mode: str = "unset") -> str:
```
Comment (gitlawr) [inline gpustack/utils/command.py]: Can `compute_mode ` be part of the `command_map`?

File: gpustack/server/controllers.py [hunk 6]
```diff
@@ -107,9 +113,15 @@ async def sync_replicas(session: AsyncSession, model: Model):
             logger.debug(f"Created model instance for model {model.name}")
 
     elif len(instances) > model.replicas:
-        for instance in instances[model.replicas :]:
-            await instance.delete(session)
-            logger.debug(f"Deleted model instance {instance.name}")
+        scheduler = Scheduler(cfg)
+        candidates = await scheduler.find_scale_down_candidates(model)
```
Comment (gitlawr) [inline gpustack/server/controllers.py]: Can we decouple controller and scheduler?

File: gpustack/worker/serve_manager.py [hunk 7]
```diff
@@ -211,3 +212,64 @@ def monitor_processes(self):
...
                 self._serving_model_instances.pop(id)
+
+    def start_rpc_servers(self):
+        try:
+            self._start_rpc_servers()
+        except Exception as e:
+            logger.error(f"Failed to start rpc servers: {e}")
+            return
+
+    def _start_rpc_servers(self):
```
Comment (gitlawr) [inline gpustack/worker/serve_manager.py]: What about we manage rpc_servers in worker_manager?

File: gpustack/schemas/models.py [hunk 8]
```diff
@@ -93,8 +105,23 @@ class ComputedResourceClaim(BaseModel):
     is_unified_memory: Optional[bool] = False
     offload_layers: Optional[int] = None
     total_layers: Optional[int] = None
-    memory: Optional[int] = Field(default=None)  # in bytes
-    gpu_memory: Optional[int] = Field(default=None)  # in bytes
+    ram: Optional[int] = Field(default=None)  # in bytes
+    vram: Optional[Dict[int, int]] = Field(default=None)  # in bytes
```
Comment (gitlawr) [inline gpustack/schemas/models.py:109]: Why do we change these? We still use gpu_memory in other APIs. Should we change as well?

Comment (aiwantaozi) [reply gpustack/schemas/models.py:109]: Others will be updated in another PR.

**Pull request merged**